app/stores/SchoolStore.js

```diff
@@ -78,117 +93,208 @@ function getBuildingForYear(school, year) {
   year = year || _getLatestYear(school);
   const schoolBuilding = getItemForYear(school.buildings, year);
   const building = (
-    schoolBuilding ? _getAssociationObject(schoolBuilding, BuildingStore.getBuilding) : {}
+    schoolBuilding ? getAssociationObject(schoolBuilding, BuildingStore.getBuilding) : {}
   );
   const schoolPrincipal = getItemForYear(school.principals, year);
   const principal = (
-    schoolPrincipal ? _getAssociationObject(schoolPrincipal, PrincipalStore.getPrincipal) : {}
+    schoolPrincipal ? getAssociationObject(schoolPrincipal, PrincipalStore.getPrincipal) : {}
   );
   const address = getItemForYear(building.addresses, year);
+  const name = getItemForYear(school.names, year) || {};
+  const id = `${school.id}-${name && name.id}-${building && building.id}-${address && address.id}`;
   return {
     address: getAddressString(address),
     archive: getItemForYear(school.archives, year) || {},
     building: building,
     gender: getItemForYear(school.genders, year) || {},
-    id: school.id,
+    id: id,
     language: getItemForYear(school.languages, year) || {},
     location: _.first(getLocationsForYear(school, year)) || {},
-    name: getItemForYear(school.names, year) || {},
+    name: name,
     principal: principal,
     type: getItemForYear(school.types, year) || {}
   };
 }
 
-function getSchoolsYearDetails(schoolIds, year) {
-  return _.map(schoolIds, function(id) {
-    return SchoolStore.getSchoolYearDetails(id, year);
-  }, this);
+function getSearchDetails(schoolIds, query) {
+  let searchDetails = [];
+  _.each(schoolIds, function(schoolId) {
+    const school = _schools[schoolId];
+    if (_.isEmpty(school)) {
+      return;
+    }
+
+    let nameSearchIndex = new Fuse(
+      school.names,
+      {keys: ['officialName']}
+    );
+    let names = nameSearchIndex.search(query);
+    names = _filterOutNamesOutsideTimelineRange(names);
+
+    _.each(names, function(name) {
+      const location = _.first(getLocationsForYear(school, name.beginYear)) || {};
+      location.id = !_.isEmpty(location) ? `${name.id}-${location.id}` : null;
+      searchDetails.push(
+        {
+          beginYear: name.beginYear,
+          endYear: name.endYear,
+          id: school.id + '-' + name.id,
+          imageUrl: getImageUrl(getBuildingForYear(school, name.beginYear)),
+          location: location,
+          name: name.officialName,
+          schoolId: school.id,
+          type: 'school-name'
+        }
+      );
+    });
+  });
+  return _.sortBy(searchDetails, 'id');
+}
+
+function getSearchDetailsForItem(schoolId, item) {
+  let searchDetails = [];
+  const school = _schools[schoolId];
+  if (_.isEmpty(school)) {
+    return [];
+  }
+
+  const names = _getNamesForTimeSpan(school, item.beginYear, item.endYear);
+  _.each(_.sortBy(names, 'beginYear'), function(name, index) {
+    const beginYear = index === 0 ? item.beginYear || name.beginYear : name.beginYear;
```

Comment:
In that case if the item.beginYear was not defined we would not have a beginYear at all. Maybe that would be preferable, but I think I was fixing some issue with that.